modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryStatistics.java

```diff
@@ -72,13 +75,14 @@
 
     /** */
     public TcpDiscoveryStatistics() {
-        joinedNodesCnt = new IntMetricImpl("JoinedNodes", "Joined nodes count");
+        joinedNodesCnt = new IntMetricImpl(metricName(DISCO_METRICS, "JoinedNodes"), "Joined nodes count");
```

Comment:
Can we make metric constructor private and provide factory method to guarantee correct creation of the metric instance?
Something like:
```
public class IntMetricImpl {
    private IntMetricImpl() {}
    public static IntMetric create(String regName, String mname, String desc) {
        return new IntMetricImpl(metricName(regName, mname), desc);
    }

    public static IntMetric create(MetricRegistry mreg, String mname, String desc) {
        return new IntMetricImpl(metricName(mreg.name(), mname), desc);
    }
}
```

Reply:
From my point of view it's better to just get rid of MetricRegistry#register(Metric) method. It is only used in case when we can't obtain metric registry but already need to collect metric value. And I think it can be replaced with MetricRegistry#register(String, IntSupplier, String) and analogs. In this case, the metrics registry will encapsulate the logic for creating and naming metrics. WDYT?

Reply:
I think the creation of some wrappers and keeping explicit `int` variables just to register metric correctly is overwhelming.